File: samples/octave/pyramid_segmentation.m
#! /usr/bin/env octave

cv;
highgui;

global g;
g.image0 = [];
g.image1 = [];
g.threshold1 = 255;
g.threshold2 = 30;
g.l = g.level = 4;
g.block_size = 1000;
g.filter = CV_GAUSSIAN_5x5;
g.storage = [];
g.min_comp = CvConnectedComp();

function set_thresh1( val )
  global g;
  g.threshold1 = val;
  ON_SEGMENT();
endfunction

function set_thresh2( val )
  global g;
  g.threshold2 = val;
  ON_SEGMENT()
endfunction

function ON_SEGMENT()
  global g;
  global cv;
  g
  swig_this(g.image0)
  swig_this(g.image1)
  swig_this(g.storage)
  g.level
  g.threshold1
  g.threshold2
  comp = cv.cvPyrSegmentation(g.image0, g.image1, g.storage, g.level, g.threshold1+1, g.threshold2+1);
  cvShowImage("Segmentation", g.image1);
endfunction

filename = "../c/fruits.jpg";
if (size(argv, 2) >= 1)
  filename = argv(){1};
endif
g.image0 = cvLoadImage( filename, 1);
if (! swig_this(g.image0))
  printf("Error opening %s\n",filename);
  exit(-1);
endif

cvNamedWindow("Source", 0);
cvShowImage("Source", g.image0);
cvNamedWindow("Segmentation", 0);
g.storage = cvCreateMemStorage ( g.block_size );
new_width = bitshift(g.image0.width, -g.level);
new_height = bitshift(g.image0.height, -g.level);
g.image0 = cvCreateImage( cvSize(new_width,new_height), g.image0.depth, g.image0.nChannels );
g.image1 = cvCreateImage( cvSize(new_width,new_height), g.image0.depth, g.image0.nChannels );
## segmentation of the color image
g.l = 1;
g.threshold1 =255;
g.threshold2 =30;
ON_SEGMENT();
g.sthreshold1 = cvCreateTrackbar("Threshold1", "Segmentation", g.threshold1, 255, @set_thresh1);
g.sthreshold2 = cvCreateTrackbar("Threshold2", "Segmentation",  g.threshold2, 255, @set_thresh2);
cvShowImage("Segmentation", image1);
cvWaitKey(0);
cvDestroyWindow("Segmentation");
cvDestroyWindow("Source");

File: samples/swig_python/pyramid_segmentation.py
#!/usr/bin/python
import sys
from opencv.cv import *
from opencv.highgui import *
image =  [None, None]
image0 = None
image1 = None
threshold1 = 255
threshold2 = 30
l = level = 4;
block_size = 1000;
filter = CV_GAUSSIAN_5x5;
storage = None
min_comp = CvConnectedComp()

def set_thresh1( val ):
    global threshold1
    threshold1 = val
    ON_SEGMENT()

def set_thresh2( val ):
    global threshold2
    threshold2 = val
    ON_SEGMENT()

def ON_SEGMENT():
    global storage
    global min_comp
    comp = cvPyrSegmentation(image0, image1, storage, level, threshold1+1, threshold2+1);
    cvShowImage("Segmentation", image1);

if __name__ == "__main__":
    filename = "../c/fruits.jpg";
    if len(sys.argv) == 2:
        filename = sys.argv[1]
    image[0] = cvLoadImage( filename, 1)
    if not image[0]:
        print "Error opening %s" % filename
        sys.exit(-1)

    cvNamedWindow("Source", 0);
    cvShowImage("Source", image[0]);
    cvNamedWindow("Segmentation", 0);
    storage = cvCreateMemStorage ( block_size );
    image[0].width &= -(1<<level);
    image[0].height &= -(1<<level);
    image0 = cvCloneImage( image[0] );
    image1 = cvCloneImage( image[0] );
    # segmentation of the color image
    l = 1;
    threshold1 =255;
    threshold2 =30;
    ON_SEGMENT();
    sthreshold1 = cvCreateTrackbar("Threshold1", "Segmentation", threshold1, 255, set_thresh1);
    sthreshold2 = cvCreateTrackbar("Threshold2", "Segmentation",  threshold2, 255, set_thresh2);
    cvShowImage("Segmentation", image1);
    cvWaitKey(0);
    cvDestroyWindow("Segmentation");
    cvDestroyWindow("Source");
